File: packages/shared-ts/src/logging/sanitize.test.ts
import { describe, expect, it } from 'bun:test';

import { sanitizeValue } from './sanitize';

describe('logging/sanitize', () => {
  describe('sanitizeValue', () => {
    it('returns non-string primitives unchanged', () => {
      expect(sanitizeValue(42)).toBe(42);
      expect(sanitizeValue(true)).toBe(true);
      expect(sanitizeValue(null)).toBe(null);
      expect(sanitizeValue(undefined)).toBe(undefined);
    });

    it('converts bigint values to strings', () => {
      expect(sanitizeValue(123n)).toBe('123');
      expect(sanitizeValue({ id: 123n })).toEqual({ id: '123' });
    });

    it('returns plain strings unchanged', () => {
      expect(sanitizeValue('hello world')).toBe('hello world');
    });

    it('redacts email addresses', () => {
      expect(sanitizeValue('contact: user@example.com')).toBe('contact: [REDACTED_EMAIL]');
    });

    it('redacts credit card numbers', () => {
      expect(sanitizeValue('card: 1234-5678-9012-3456')).toBe('card: [REDACTED_CREDIT_CARD]');
      expect(sanitizeValue('card: 1234 5678 9012 3456')).toBe('card: [REDACTED_CREDIT_CARD]');
    });

    it('redacts API keys', () => {
      expect(sanitizeValue('key: sk_1234567890abcdefghijklmnop')).toBe('key: [REDACTED_API_KEY]');
      expect(sanitizeValue('key: api_1234567890abcdefghijklmnop')).toBe('key: [REDACTED_API_KEY]');
    });

    it('redacts Stripe keys', () => {
      expect(sanitizeValue('sk_test_1234567890abcdefghijklmnop')).toBe('[REDACTED_STRIPE_KEY]');
      expect(sanitizeValue('pk_live_1234567890abcdefghijklmnop')).toBe('[REDACTED_STRIPE_KEY]');
    });

    it('redacts SSN', () => {
      expect(sanitizeValue('ssn: 123-45-6789')).toBe('ssn: [REDACTED_SSN]');
    });

    it('redacts phone numbers', () => {
      expect(sanitizeValue('phone: 123-456-7890')).toBe('phone: [REDACTED_PHONE]');
      expect(sanitizeValue('phone: 123.456.7890')).toBe('phone: [REDACTED_PHONE]');
    });

    it('redacts JWT tokens', () => {
      const jwt =
        'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIn0.dozjgNryP4J3jVmNHl0w5N_XgL0n3I9PlFUP0THsR8U';
      expect(sanitizeValue(`token: ${jwt}`)).toBe('token: [REDACTED_JWT]');
    });

    it('redacts Bearer tokens', () => {
      expect(sanitizeValue('Authorization: Bearer abc123xyz')).toBe(
        'Authorization: [REDACTED_BEARER_TOKEN]'
      );
      expect(sanitizeValue('Authorization: bEaReR abc123xyz')).toBe(
        'Authorization: [REDACTED_BEARER_TOKEN]'
      );
    });

    it('redacts dotted Bearer tokens', () => {
      expect(sanitizeValue('Authorization: Bearer abc.def.ghi')).toBe(
        'Authorization: [REDACTED_BEARER_TOKEN]'
      );
    });

    it('handles arrays recursively', () => {
      expect(sanitizeValue(['hello', 'user@example.com', 42])).toEqual([
        'hello',
        '[REDACTED_EMAIL]',
        42,
      ]);
    });

    it('redacts keys named apikey or api_key', () => {
      expect(sanitizeValue({ apiKey: 'secret123' })).toEqual({
        apiKey: '[REDACTED_API_KEY]',
      });
      expect(sanitizeValue({ api_key: 'secret123' })).toEqual({
        api_key: '[REDACTED_API_KEY]',
      });
    });

    it('redacts password, secret, and token keys', () => {
      expect(sanitizeValue({ password: 'secret123' })).toEqual({
        password: '[REDACTED]',
      });
      expect(sanitizeValue({ secret: 'my-secret' })).toEqual({
        secret: '[REDACTED]',
      });
      expect(sanitizeValue({ authToken: 'abc123' })).toEqual({
        authToken: '[REDACTED]',
      });
    });

    it('recursively sanitizes nested objects', () => {
      expect(
        sanitizeValue({
          user: {
            email: 'user@example.com',
            name: 'John',
          },
          credentials: {
            password: 'secret',
          },
        })
      ).toEqual({
        user: {
          email: '[REDACTED_EMAIL]',
          name: 'John',
        },
        credentials: {
          password: '[REDACTED]',
        },
      });
    });

    it('handles mixed nested structures', () => {
      expect(
        sanitizeValue({
          items: ['user@test.com', { apiKey: 'key123' }],
          value: 123,
        })
      ).toEqual({
        items: ['[REDACTED_EMAIL]', { apiKey: '[REDACTED_API_KEY]' }],
        value: 123,
      });
    });

    it('handles circular references without throwing', () => {
      const circular: Record<string, unknown> = { email: 'user@example.com' };
      circular['self'] = circular;

      expect(sanitizeValue(circular)).toEqual({
        email: '[REDACTED_EMAIL]',
        self: '[Circular]',
      });
    });

    it('handles circular arrays without throwing', () => {
      const circular: unknown[] = ['user@example.com'];
      circular.push(circular);

      expect(sanitizeValue(circular)).toEqual(['[REDACTED_EMAIL]', '[Circular]']);
    });
  });
});

File: packages/core/go/pkg/redaction/sanitize_test.go
package redaction

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func fakeStripeKey(environment string) string {
	return "sk_" + environment + "_abcdefghijklmnopqrstuvwxyz"
}

func TestSanitizeValue(t *testing.T) {
	tests := []struct {
		name     string
		input    any
		expected any
	}{
		{
			name:     "String",
			input:    "hello world",
			expected: "hello world",
		},
		{
			name:     "Email",
			input:    "Contact: test@example.com",
			expected: "Contact: [REDACTED_EMAIL]",
		},
		{
			name:     "API Key",
			input:    "Key: " + fakeStripeKey("live"),
			expected: "Key: [REDACTED_STRIPE_KEY]",
		},
		{
			name:     "Map with password",
			input:    map[string]any{"password": "secret123", "username": "user"},
			expected: map[string]any{"password": "[REDACTED]", "username": "user"},
		},
		{
			name:     "Map with api key key",
			input:    map[string]any{"my_api_key": "12345"},
			expected: map[string]any{"my_api_key": "[REDACTED_API_KEY]"},
		},
		{
			name:     "Nested Map",
			input:    map[string]any{"user": map[string]any{"email": "test@test.com"}},
			expected: map[string]any{"user": map[string]any{"email": "[REDACTED_EMAIL]"}},
		},
		{
			name:     "Slice",
			input:    []any{"safe", "test@test.com"},
			expected: []any{"safe", "[REDACTED_EMAIL]"},
		},
		{
			name:     "Typed Slice",
			input:    []string{"safe", "test@test.com"},
			expected: []any{"safe", "[REDACTED_EMAIL]"},
		},
		{
			name:     "Map with secret",
			input:    map[string]any{"client_secret": "abc123"},
			expected: map[string]any{"client_secret": "[REDACTED]"},
		},
		{
			name:     "Map with token",
			input:    map[string]any{"access_token": "xyz789"},
			expected: map[string]any{"access_token": "[REDACTED]"},
		},
		{
			name:     "Map with apikey (no underscore)",
			input:    map[string]any{"myapikey": "key123"},
			expected: map[string]any{"myapikey": "[REDACTED_API_KEY]"},
		},
		{
			name: "JWT",
			input: "token: eyJhbGciOiJIUzI1NiJ9." +
				"eyJzdWIiOiIxMjM0NTY3ODkwIn0.signature_value-123",
			expected: "token: [REDACTED_JWT]",
		},
		{
			name: "Bearer token",
			input: "Authorization: Bearer " +
				"eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIxIn0.signature_value-123",
			expected: "Authorization: [REDACTED_BEARER_TOKEN]",
		},
		{
			name:     "Bearer token with tab separator",
			input:    "Authorization: Bearer\tsecret-token-value",
			expected: "Authorization: [REDACTED_BEARER_TOKEN]",
		},
		{
			name:     "Short email",
			input:    "a@b.co",
			expected: "[REDACTED_EMAIL]",
		},
		{
			name:     "Non-string non-map non-slice value",
			input:    12345,
			expected: 12345,
		},
		{
			name:     "nil value",
			input:    nil,
			expected: nil,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			assert.Equal(t, tt.expected, SanitizeValue(tt.input))
		})
	}
}

func TestSanitizeValueAdditionalPatterns(t *testing.T) {
	assert.Equal(t, "[REDACTED_CREDIT_CARD]", SanitizeValue("4111111111111111"))
	assert.Equal(t, "[REDACTED_SSN]", SanitizeValue("123-45-6789"))
	assert.Equal(t, "[REDACTED_PHONE]", SanitizeValue("555-123-4567"))
	assert.Equal(t, "[REDACTED_API_KEY]", SanitizeValue("tfai_1234567890abcdef1234567890abcdef12345678"))
	assert.Equal(t, "[REDACTED_API_KEY]", SanitizeValue("sk_1234567890abcdef1234567890abcdef"))
	assert.Equal(t, "[REDACTED_API_KEY]", SanitizeValue("pk_1234567890abcdef1234567890abcdef"))
	assert.Equal(t, "[REDACTED_API_KEY]", SanitizeValue("api_abcdefghijklmnopqrstuvwxyz"))
	assert.Equal(t, "[REDACTED_API_KEY]", SanitizeValue("KEY_abcdefghijklmnopqrstuvwxyz"))
	assert.Equal(t, "[REDACTED_API_KEY]", SanitizeValue("token_abcdefghijklmnopqrstuvwxyz"))
	assert.Equal(t, map[string]any{"items": []any{"[REDACTED_EMAIL]"}}, SanitizeValue(map[string]any{
		"items": []any{"person@example.com"},
	}))
}

func TestSanitizeValueReflectPaths(t *testing.T) {
	type tokenPayload struct {
		Token string
		Email string
	}

	assert.Equal(t, []byte{1, 2, 3}, SanitizeValue([]byte{1, 2, 3}))
	assert.Equal(t, []any{"safe", "[REDACTED_EMAIL]"}, SanitizeValue([]string{"safe", "person@example.com"}))
	assert.Equal(t, map[int]string{1: "x"}, SanitizeValue(map[int]string{1: "x"}))
	assert.Equal(t, map[string]any{
		"Token": "[REDACTED]",
		"Email": "[REDACTED_EMAIL]",
	}, SanitizeValue(tokenPayload{
		Token: "secret-token",
		Email: "person@example.com",
	}))
}

func TestSanitizeValuePointersArraysAndNilCollections(t *testing.T) {
	type tokenPayload struct {
		Secret string
		Next   *tokenPayload
	}

	email := "person@example.com"
	assert.Equal(t, "[REDACTED_EMAIL]", SanitizeValue(&email))

	var nilPointer *string
	assert.Equal(t, nilPointer, SanitizeValue(nilPointer))

	assert.Equal(t, []any{"safe", "[REDACTED_EMAIL]"}, SanitizeValue([2]string{"safe", "person@example.com"}))

	var nilSlice []string
	assert.Equal(t, []any{}, SanitizeValue(nilSlice))

	payload := &tokenPayload{Secret: "super-secret"}
	payload.Next = payload
	assert.Equal(t, map[string]any{
		"Secret": "[REDACTED]",
		"Next":   "[Circular]",
	}, SanitizeValue(payload))
}

func TestSanitizeValueCircularReferences(t *testing.T) {
	m := map[string]any{"email": "person@example.com"}
	m["self"] = m

	assert.Equal(t, map[string]any{
		"email": "[REDACTED_EMAIL]",
		"self":  "[Circular]",
	}, SanitizeValue(m))

	arr := make([]any, 2)
	arr[0] = "safe"
	arr[1] = arr
	assert.Equal(t, []any{"safe", "[Circular]"}, SanitizeValue(arr))
}

var benchmarkSanitizedValue any

func BenchmarkSanitizeValueSafeString(b *testing.B) {
	for b.Loop() {
		benchmarkSanitizedValue = SanitizeValue("GET /api/v1/tasks completed")
	}
}

func BenchmarkSanitizeValueSensitiveString(b *testing.B) {
	for b.Loop() {
		benchmarkSanitizedValue = SanitizeValue("Authorization: Bearer eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIxIn0.signature_value-123")
	}
}

func BenchmarkSanitizeValueSafeMetadata(b *testing.B) {
	metadata := map[string]any{
		"method": "GET",
		"path":   "/api/v1/tasks",
		"status": 200,
		"tags":   []any{"engine", "stream", "completed"},
		"nested": map[string]any{
			"task_id": "task-123",
			"model":   "openai/gpt-5",
		},
	}

	b.ReportAllocs()
	for b.Loop() {
		benchmarkSanitizedValue = SanitizeValue(metadata)
	}
}

func BenchmarkSanitizeValueSensitiveMetadata(b *testing.B) {
	metadata := map[string]any{
		"email":        "person@example.com",
		"access_token": "secret",
		"headers": map[string]any{
			"authorization": "Bearer eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIxIn0.signature_value-123",
		},
		"events": []any{
			map[string]any{"message": "contact admin@example.com"},
			map[string]any{"status": "complete"},
		},
	}

	b.ReportAllocs()
	for b.Loop() {
		benchmarkSanitizedValue = SanitizeValue(metadata)
	}
}
